From 0432e2519d524d6c2a77e199cf4b438c8a5a5cdf Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Tue, 18 Mar 2025 10:59:52 +0100
Subject: [PATCH 822/954] feat(st): drivers/stm32_iwdg: set watchdog timeout
 from DT

Read the timeout value from DT and program the watchdog.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 389bf6b22206e3597f040f16801697e41f6b69a0 feat(st): drivers/stm32_iwdg: set watchdog timeout from DT]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I15b3ded2bc1a2163b768fe3ff71c5aa17eebaafc
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/444806
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/iwdg/stm32_iwdg.c | 119 +++++++++++++++++++++++++++++++----
 1 file changed, 107 insertions(+), 12 deletions(-)

diff --git a/drivers/st/iwdg/stm32_iwdg.c b/drivers/st/iwdg/stm32_iwdg.c
index f01660af66..6f4660a8f7 100644
--- a/drivers/st/iwdg/stm32_iwdg.c
+++ b/drivers/st/iwdg/stm32_iwdg.c
@@ -25,18 +25,64 @@
 
 /* IWDG registers offsets */
 #define IWDG_KR_OFFSET		0x00U
+#define IWDG_PR_OFFSET		0x04U
+#define IWDG_RLR_OFFSET		0x08U
 
 /* Registers values */
-#define IWDG_KR_RELOAD_KEY	0xAAAA
-#define IWDG_KR_ENABLE_KEY	0xCCCC
+#define IWDG_KR_EN_WRITE_KEY	0x5555U
+#define IWDG_KR_RELOAD_KEY	0xAAAAU
+#define IWDG_KR_ENABLE_KEY	0xCCCCU
+
+#define IWDG_PRESCALER_MIN	2U
+#define IWDG_PRESCALER_MAX	10U
+#define IWDG_RLR_BITS		12U
 
 struct stm32_iwdg_instance {
 	uintptr_t base;
-	unsigned long clock;
+	unsigned long clk_lsi;
+	unsigned long clk_pclk;
+	unsigned long timeout;
 };
 
 static struct stm32_iwdg_instance stm32_iwdg[IWDG_MAX_INSTANCE];
 
+static unsigned int ilog2(unsigned int val)
+{
+	unsigned int res = 0U;
+
+	while (val > 1U) {
+		val >>= 1U;
+		res++;
+	}
+
+	return res;
+}
+
+static int fdt_get_clk_by_name(const void *dtb, int node, const char *name,
+			       unsigned long *clk)
+{
+	const fdt32_t *cuint;
+	int index, len;
+
+	index = fdt_stringlist_search(dtb, node, "clock-names", name);
+	if (index < 0) {
+		return index;
+	}
+
+	cuint = fdt_getprop(dtb, node, "clocks", &len);
+	if (cuint == NULL) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	if (len < (2 * (index + 1) * (int)sizeof(uint32_t))) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	*clk = fdt32_to_cpu(cuint[(2 * index) + 1]);
+
+	return 0;
+}
+
 static int stm32_iwdg_get_dt_node(struct dt_node_info *info, int offset)
 {
 	int node;
@@ -61,28 +107,60 @@ void stm32_iwdg_refresh(void)
 
 		/* 0x00000000 is not a valid address for IWDG peripherals */
 		if (iwdg->base != 0U) {
-			clk_enable(iwdg->clock);
+			clk_enable(iwdg->clk_pclk);
 
 			mmio_write_32(iwdg->base + IWDG_KR_OFFSET,
 				      IWDG_KR_RELOAD_KEY);
 
-			clk_disable(iwdg->clock);
+			clk_disable(iwdg->clk_pclk);
 		}
 	}
 }
 
 static void stm32_iwdg_start(struct stm32_iwdg_instance *iwdg)
 {
+	unsigned long lsi_rate;
+	unsigned long long reload;
+	uint32_t rlr, pr;
+
 	/* 0x00000000 is not a valid address for IWDG peripherals */
 	if (iwdg->base == 0U) {
 		panic();
 	}
 
-	clk_enable(iwdg->clock);
+	lsi_rate = clk_get_rate(iwdg->clk_lsi);
+	if (lsi_rate == 0U) {
+		panic();
+	}
+
+	reload = iwdg->timeout * (unsigned long long)lsi_rate;
+	if ((reload >> (IWDG_RLR_BITS + IWDG_PRESCALER_MAX)) != 0ULL) {
+		/* use max timeout */
+		pr = IWDG_PRESCALER_MAX - IWDG_PRESCALER_MIN;
+		rlr = GENMASK_32(IWDG_RLR_BITS - 1U, 0U);
+	} else if ((reload >> IWDG_PRESCALER_MIN) <= 2ULL) {
+		/* Be safe and expect any counter to be above 2 */
+		panic();
+	} else {
+		unsigned int bits = ilog2((unsigned int)reload) + 1U;
 
+		if (bits < (IWDG_RLR_BITS + IWDG_PRESCALER_MIN)) {
+			pr = 0U;
+		} else {
+			pr = bits - IWDG_RLR_BITS - IWDG_PRESCALER_MIN;
+		}
+
+		rlr = (uint32_t)(reload >> (pr + IWDG_PRESCALER_MIN)) - 1U;
+	}
+
+	clk_enable(iwdg->clk_pclk);
+
+	mmio_write_32(iwdg->base + IWDG_KR_OFFSET, IWDG_KR_EN_WRITE_KEY);
+	mmio_write_32(iwdg->base + IWDG_PR_OFFSET, pr);
+	mmio_write_32(iwdg->base + IWDG_RLR_OFFSET, rlr);
 	mmio_write_32(iwdg->base + IWDG_KR_OFFSET, IWDG_KR_ENABLE_KEY);
 
-	clk_disable(iwdg->clock);
+	clk_disable(iwdg->clk_pclk);
 }
 
 int stm32_iwdg_init(void)
@@ -100,15 +178,16 @@ int stm32_iwdg_init(void)
 	     node != -FDT_ERR_NOTFOUND;
 	     node = stm32_iwdg_get_dt_node(&dt_info, node)) {
 		struct stm32_iwdg_instance *iwdg;
+		const fdt32_t *cuint;
 		uint32_t hw_init;
 		uint32_t idx;
+		int ret;
 
 		count++;
 
 		idx = stm32_iwdg_get_instance(dt_info.base);
 		iwdg = &stm32_iwdg[idx];
 		iwdg->base = dt_info.base;
-		iwdg->clock = (unsigned long)dt_info.clock;
 
 		/* Read OTP to know if IWDG is started by hardware or not */
 		hw_init = stm32_iwdg_get_otp_config(idx);
@@ -127,6 +206,24 @@ int stm32_iwdg_init(void)
 			continue;
 		}
 
+		ret = fdt_get_clk_by_name(fdt, node, "pclk", &iwdg->clk_pclk);
+		if (ret != 0) {
+			return ret;
+		}
+
+		ret = fdt_get_clk_by_name(fdt, node, "lsi", &iwdg->clk_lsi);
+		if (ret != 0) {
+			return ret;
+		}
+
+		cuint = fdt_getprop(fdt, node, "timeout-sec", NULL);
+		if (cuint == NULL) {
+			ERROR("iwdg%u: missing timeout-sec property",
+			      idx + 1U);
+			panic();
+		}
+		iwdg->timeout = fdt32_to_cpu(*cuint);
+
 		VERBOSE("IWDG%u found, %ssecure\n", idx + 1U,
 			((dt_info.status & DT_NON_SECURE) != 0) ?
 			"non-" : "");
@@ -137,10 +234,8 @@ int stm32_iwdg_init(void)
 			stm32mp_register_secure_periph_iomem(iwdg->base);
 		}
 
-		if ((hw_init & IWDG_HW_ENABLED) == 0) {
-			/* Use default timeout, ignore DT's "timeout-sec" */
-			stm32_iwdg_start(iwdg);
-		}
+		clk_enable(iwdg->clk_lsi);
+		stm32_iwdg_start(iwdg);
 	}
 
 	VERBOSE("%u IWDG instance%s found\n", count, (count > 1U) ? "s" : "");
-- 
2.53.0

